droid-hal-device.inc

```diff
@@ -829,20 +785,62 @@ fi
 %endif
 %endif
 
+%pre users
+#Compatibility with old droid-hal version - convert previous format
+#of droid-specific user/group information to current one.
+cd %{_libdir}/droid
+if [ $1 -gt 1 -a \
+     -e droid-user-remove.sh.installed ]; then
+  echo '#!/bin/sh' > droid-user-remove.sh.installed
+fi
+
+%post users
+cd %{_libdir}/droid
+
+# NOTE: We do not delete groups/users in any case since that would cause issues
+# with users being removed from groups where they should really be in.
+
+echo creating droid users and groups
+[ -e droid.ids ] && cat droid.ids | while read name id; do
+    if ! getent group $name; then
+        echo "Group $name did not exist yet"
+        groupadd -g $id $name || :
+    else
+        echo "Group $name already existed"
+        groupmod -g $id $name || :
+    fi
+    if ! getent passwd $name; then
```

Comment:
With the current shape this PR makes a regression where it's no longer supported for the android adaptation to change names and/or IDs of android user/groups.

The pre-PR code handled this by removing all users/groups it created, including the group members. My code tried to avoid the situation by assuming that the package is never removed, just updated. In that case, the users/group whose IDs did not change were not touched, and thus the membership added by other packages was retained. At the same time if an android ID was modified, the change was propagated, and the user/group was removed and re-created with a proper ID. Since it was removed, if any other packages added members to that renumbered group, the membership was lost.

The modified code makes a convenient assumption that we do not need to remove users/groups when the package is removed. After all, some stray user/group that no longer exists in the new Android adaptation should not cause us any issues by staying in the system. The only case where I see it might is if it's got an ID that is re-used in a different Android user/group or in Sailfish.

So, the real problem I see here is if the android IDs change:
- if the name is changed and ID stays the same, we either have 2 user/groups with the same ID or usermod below fails (I didn't check if it would).
- if the ID changes and the name stays the same, this getent will be glad to find the user, and will not add the new, correct one.

My gut feeling tells me that the android IDs should rarely change, if at all. Additions of new entries is supported, so even if there's a new one, it's ok. At the time I was writing the PR I was persuaded that android ID changes actually happen. Still, I think that was probably during early device adaptation process, and I don't think we bumped the android adaptation layer in any device so significantly as to change android IDs. 

Following my gut feeling, the most reasonable thing to do is to have the package create the users/groups and never touch them afterwards. When there's a change in the android adaptation layer, one needs to verify what change it is, and if there's any ID removal or rename, handling for that case needs to be added to the particular device's droid-hal-$device-users instead of the common, generic package that would then only handle additions.

Reply:
Comment on the problems you mentioned:
1. If the name is changed and ID stays the same, default useradd/groupadd will not allow duplicate gid, but it can be forced. Though I'd argue we shouldn't have duplicate gids.
2. If the ID changes and the name stays the same, current PR would change the uid/gid to new one, so that's not a problem.

From those issues first one is something that *could* be an issue. If so, for that case it might make sense to implement something where the old group is removed, new one created and users from previous group are re-added to the new one.

Reply:
2) is a problem if the group is shared by SFOS and files in the filesystem already exist under that group, i.e. after changing the ID things would loose access to those files
we had this case already once with the dbus group, (android uses a different id for dbus) so we had to rename the android dbus group to adbus.

1) i think @jusa's suggestion for handling this issue could be done but that would also be handled specially if the group is shared with SFOS, since we should not remove SFOS groups at all.

Reply:
> With the current shape this PR makes a regression where it's no longer supported for the android adaptation to change names and/or IDs of android user/groups.

On the contrary, with the change group names and/or gids are updated to the new ones if there are changes. However, like you said, if a group is not used in new version it is not removed but having unused groups is less problematic than removing them.

Reply:
Indeed, I must have overlooked the else statement, that's why my understanding of the code was different from what happens as per @jusa's point 2.

With the updates by @krnlyng, I think the right thing, "best effort" happens. If we find a duplicate ID, we rename the group, if we find a duplicate name, we renumber that group.